docs/architecture/adr-versions.md

```diff
@@ -0,0 +1,219 @@
+- It's easy to determine the version of a block from its serialized form
+
+#### P2PVersion
+
+- All peer messaging (messages, detectable behaviour)
+- Will change gradually as reactors evolve to improve performance and support new features
+	- eg proposed new message types BatchTx in the mempool and HasBlockPart in the consensus
+- It's easy to determine the version of a peer from its first serialized message/s
+- New versions must be compatible with at least one old version to allow gradual upgrades
+
+#### AppVersion
+
+- The ABCI state machine (txs, begin/endblock behaviour, commit hashing)
+- Behaviour and message types will change abruptly in the course of the life of a chain
+- Need to minimize complexity of the code for supporting different AppVersions at different heights
+- Ideally, each version of the software supports only a *single* AppVersion at one time
+    - this means we checkout different versions of the software at different heights instead of littering the code
+          with conditionals
+    - minimize the number of data migrations required across AppVersion (ie. most AppVersion should be able to read the same state from disk as previous AppVersion).
+- NOTE: we still need to support peers syncing from old heights across AppVersion changes
+
+### Networks
+
+We need to support many independent networks running the same version of the software,
+even possibly starting from the same initial state.
+They must have distinct identifiers so that peers know which one they are joining and so
+validators and users can prevent replay attacks.
+
+Call this the `NetworkName`. It represents both the application being run and the community or intention
+of running it.
+
+Peers only connect to other peers with the same NetworkName.
+
+### Forks
+
+We need to support existing networks upgrading and forking, wherein they may do any of:
+	- revert back to some height, continue with the same versions but new blocks
+	- arbitrarily mutate state at some height, continue with the same versions (eg. Dao Fork)
+	- change the AppVersion at some height
+
+Note because of Tendermint's voting power threshold rules, a chain can only be extended under the "original" rules and under the new rules
+if 1/3 or more is double signing, which is expressly prohibited, and is supposed to result in their punishment on both chains. Since they can censor
+the punishment, the chain is expected to be hardforked to remove the validators. Thus, if both branches are to continue after a fork,
+they will each require a new identifier, and the old chain identifier will be retired (ie. only useful for syncing history, not for new blocks)..
+
+We need a consistent way to describe forks.
+
+## Ideal
+
+Each component of the software is independently versioned in a modular way and its easy to mix and match and upgrade.
+
+Good luck pal ;)
+
+## Proposal
+
+### Versions
+
+Each of BlockVersion, AppVersion, P2PVersion is a monotonically increasing int64.
```

Comment:
It seems that we assume that versions will never be backward compatible, i.e., that every new version will be breaking. Why we don't we use semantic versioning for versions? At least for P2PVersion we could probably have protocol change that is backward compatible (for example performance improvement that adds new message types, but that can still work with the old versions).

Reply:
Semantic version doesn't cut it - we want to be able to make a sequence of non-breaking changes that result in a breaking change. ie. v20 is compatible with v19, and v21 is compatible with v20, but v21 is not compatible with v19.

This proposal is assuming for simplicity that the Block and App protocol versions are never backwards compatible